// jparticles/lib/common/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENT_NAMES_SNOW = exports.EVENT_NAMES_WAVE_LOADING = exports.EVENT_NAMES = exports.regExp = exports.isRuntimeSupported = exports.isIE8 = exports.defaultCanvasHeight = exports.defaultCanvasWidth = exports.objectCtorString = exports.funcToString = exports.orientationSupport = exports.doublePi = exports.piBy180 = void 0;
exports.piBy180 = Math.PI / 180;
exports.doublePi = Math.PI * 2;
exports.orientationSupport = !!window.DeviceOrientationEvent;
exports.funcToString = Function.prototype.toString;
exports.objectCtorString = exports.funcToString.call(Object);
exports.defaultCanvasWidth = 485;
exports.defaultCanvasHeight = 300;
exports.isIE8 = /msie\s8.0/i.test(navigator.userAgent);
exports.isRuntimeSupported = !!Object.defineProperty && !exports.isIE8;
// 正则列表
exports.regExp = {
    // 匹配任意空白
    trimAll: /\s/g,
    // 匹配 HTTP 协议
    http: /^(https?:\/\/|\/\/)/i,
    // 匹配 'star' 或 'star:5:0.5'
    shapeStar: /^star(:\d+:\d+(\.\d+)?)?$/,
    // 匹配 Base64 图片
    imageBase64: /^data:image\/(png|jpe?g|gif|svg\+xml);base64,/,
};
// 公共事件名列表
var EVENT_NAMES;
(function (EVENT_NAMES) {
    EVENT_NAMES["DESTROY"] = "DESTROY";
    EVENT_NAMES["RESIZE"] = "RESIZE";
})(EVENT_NAMES = exports.EVENT_NAMES || (exports.EVENT_NAMES = {}));
// WaveLoading 事件名
var EVENT_NAMES_WAVE_LOADING;
(function (EVENT_NAMES_WAVE_LOADING) {
    EVENT_NAMES_WAVE_LOADING["PROGRESS"] = "PROGRESS";
    EVENT_NAMES_WAVE_LOADING["FINISHED"] = "FINISHED";
})(EVENT_NAMES_WAVE_LOADING = exports.EVENT_NAMES_WAVE_LOADING || (exports.EVENT_NAMES_WAVE_LOADING = {}));
// Snow 事件名
var EVENT_NAMES_SNOW;
(function (EVENT_NAMES_SNOW) {
    EVENT_NAMES_SNOW["FINISHED"] = "FINISHED";
})(EVENT_NAMES_SNOW = exports.EVENT_NAMES_SNOW || (exports.EVENT_NAMES_SNOW = {}));
